src/consensus/aft/test/committable_suffix.cpp

```diff
@@ -0,0 +1,860 @@
+  std::optional<size_t> max_to_keep = std::nullopt)
+{
+  auto it = messages.begin();
+  std::map<ccf::NodeId, size_t> kept;
+  while (it != messages.end())
+  {
+    if (
+      std::find(targets.begin(), targets.end(), it->first) == targets.end() ||
+      (max_to_keep.has_value() && kept[it->first] >= *max_to_keep))
+    {
+      it = messages.erase(it);
+    }
+    else
+    {
+      ++it;
+      ++kept[it->first];
+    }
+  }
+}
+
+void keep_messages_for(
+  const ccf::NodeId& target,
+  aft::ChannelStubProxy::MessageList& messages,
+  std::optional<size_t> max_to_keep = std::nullopt)
+{
+  keep_messages_for_multiple({target}, messages, max_to_keep);
+}
+
+void keep_first_for(
+  const ccf::NodeId& target,
+  aft::ChannelStubProxy::MessageList& messages,
+  std::optional<size_t> max_to_keep = std::nullopt)
+{
+  keep_messages_for(target, messages, 1);
+}
+
+void keep_earliest_append_entries_for_each_target(
+  aft::ChannelStubProxy::MessageList& messages)
+{
+  std::map<
+    ccf::NodeId,
+    std::pair<aft::Index, aft::ChannelStubProxy::MessageList::iterator>>
+    best;
+  for (auto it = messages.begin(); it != messages.end(); ++it)
+  {
+    const auto& [target, contents] = *it;
+
+    const uint8_t* data = contents.data();
+    auto size = contents.size();
+    auto msg_type = serialized::peek<aft::RaftMsgType>(data, size);
+    if (msg_type == aft::raft_append_entries)
+    {
+      const auto ae = *(aft::AppendEntries*)data;
+
+      const auto best_it = best.find(target);
+      if (best_it == best.end() || best_it->second.first > ae.prev_idx)
+      {
+        best[target] = std::make_pair(ae.prev_idx, it);
+      }
+    }
+  }
+
+  aft::ChannelStubProxy::MessageList best_only;
+  for (const auto& [node_id, pair] : best)
+  {
+    const auto& [idx, it] = pair;
+    best_only.push_back(*it);
+  }
+
+  messages = std::move(best_only);
+}
+
+#define TEST_DECLARE_NODE(N) \
+  ccf::NodeId node_id##N(#N); \
+  auto store##N = std::make_shared<AllSigsStore>(node_id##N); \
+  TRaft r##N( \
+    ConsensusType::CFT, \
+    std::make_unique<AllSigsAdaptor>(store##N), \
+    std::make_unique<aft::LedgerStubProxy>(node_id##N), \
+    std::make_shared<aft::ChannelStubProxy>(), \
+    std::make_shared<aft::StubSnapshotter>(), \
+    nullptr, \
+    nullptr, \
+    cert, \
+    std::make_shared<aft::State>(node_id##N), \
+    nullptr, \
+    nullptr, \
+    nullptr, \
+    request_timeout, \
+    election_timeout, \
+    election_timeout); \
+  initial_config[node_id##N] = {}; \
+  nodes[node_id##N] = &r##N; \
+  auto channels##N = channel_stub_proxy(r##N);
+
+/**
+  Summary of the behaviour this test aims to explore (for implementation reasons
+  the indices in the actual test don't match those in this comment, but the
+  logical relationships are the same):
+
+  - Produce an initial state where A is primary in term 1 of a 5-node network,
+    has mixed success replicating its entries. Local commit index is marked by
```

Comment:
```suggestion
    has mixed success replicating its entries. Local view of the commit index is marked by
```

We've used (and continue to use I believe) "local commit index" with very different semantics.

Reply:
```suggestion
    has mixed success replicating its entries. Each node's commit index is marked by
```